### DISCUSSION TITLE
React Native Support

### DISCUSSION THREAD

**Discussion by stanlemon** [jsforce/jsforce#1465] [Ideas]
First off: Thank you so much for an awesome project. I've used this several times for testing integrations and doing demos, it's a really great library for working with Salesforce.

I have opened up two PRs that make jsforce function correctly with React Native:
- #571 Fixed JSONP detection by additionally checking for ```document```, this is important because in RN JSONP should not be used, but because the ```window``` object exists, but ```document``` does not.  This additional check should not present any regression to the existing JSONP support.
- #572 Switches to using webpack.  This one I am less confident about, but as far as I can tell things seem functional.  Why do this? First off it's important to note that RN does **not** have node core libraries, so some process must be used to load (or browserify) the various core node libraries jsforce requires.  However, RN does have a require() method. Ultimately this comes down to the way that browserify attempts to resolve dependencies using require(), which is different from what webpack does.  Webpack does not use require() but __webpack_require__() which makes all the world of difference.  If there was a way to change how browserify does this, I would prefer that, but as I am not aware of one I took a stab at converting to webpack instead.

With the two above changes it is possible to use jsforce in a React Native application by simply doing:
```
import jsforce from 'jsforce/build/jsforce.js';
```

I understand the webpack change may be undesirable, but please give it some consideration as I think being able to use jsforce in React Native is a killer feature.

**Comment by stomita** [comment]
@stanlemon Interesting. I personally am not using React Native, however, I have a will to support the env as much as possible.

From curiosity I'd like to ask you why you are importing the built file, not the module name like following.
```javascript
import jsforce from "jsforce";
```

**Comment by stanlemon** [comment]
Great question @stomita, the reason is because React Native is **not** a nodejs environment, so the built libraries like ```http``` or ```query string``` do not exist. Everything in this file https://github.com/substack/node-browserify/blob/master/lib/builtins.js needs to be accounted for.  Some have the same name as their native counterparts, ie. ```events``` but others like ```domain``` are supported through ```domain-browser```.  When jsforce does a ```require('domain');``` without a pre-processor like browserify or webpack it does not know that it should include ```domain-browser``` instead.  Ergo why a built, or processed version of jsforce is needed inside of React Native.  This is unfortunately a really common problem when integrating libraries into a React Native project.

**Comment by chrisl8888** [comment]
Are there any interest from the maintainers to move toward using es module spec instead of commonjs spec?

As @stanlemon indicates, it would improve interpolarity with client-side libraries, like react, angular, etc.

**Comment by stomita** [comment]
> Are there any interest from the maintainers to move toward using es module spec instead of commonjs spec?

@chrisjlee It would be in 2.0.
#593 
There is not such a progress to be shared here, though.

**Comment by fxfactorial** [comment]
RN is fine with commonJS and the querying stuff like http can be replaced with fetch.

**Comment by darasus** [comment]
Any updates on this issue?

**Comment by shuzootani** [comment]
@stomita Any updates on this? otherwise I should probably use mobile sdk which is tricky and inflexible to configure.

**Comment by AkshayCloudAnalogy** [comment]
Any update on this?

**Comment by stomita** [comment]
@AkshayCloudAnalogy You can try ver 2.0 alpha by`npm i jsforce@alpha` and hopefully can be used in RN (not tested, though). If there is problem please file issues.

**Comment by stanlemon** [comment]
Using expo (the recommended way to start a new react-native project), a vanilla project, and jsforce@alpha it breaks due to the 'events' module not being available.

<img width="403" alt="Screen Shot 2020-01-07 at 2 31 43 PM" src="https://user-images.githubusercontent.com/86314/71923191-7223fc00-315a-11ea-992f-afb80e4a8d8f.png">

**Comment by NishaLodhi** [comment]
Any update on this ?

**Comment by bilal-elchami** [comment]
Any new updates on this ?

**Comment by johnpanos** [comment]
Any updates on progress? We would like to create a new app for our Org that shares code between React and React Native...

**Comment by hugo-chq** [comment]
we've had some success using this (v1.x) with [node-libs-react-native](https://github.com/parshap/node-libs-react-native) and [react-native-app-auth](https://github.com/FormidableLabs/react-native-app-auth)

**Comment by pedrogmbh** [comment]
Hi, just wanted to point out the issue still exists while trying to use RN with JSForce.
